Add unit dropdown

maryx opened the pull request:
- Sets up the interactive (not static) unit converter (https://github.com/flutter/udacity-course/issues/3)
- Adds grid view for each Unit Category. Tapping on each grid item routes you to a converter page.
- Adds Text Input and Dropdown for each unit (https://github.com/flutter/udacity-course/issues/7)

Future PRs will update the UI and do the conversion math.

![grid](https://user-images.githubusercontent.com/3357543/31028451-e336c364-a502-11e7-8d69-2f2d0a3fafe1.png)
![converter](https://user-images.githubusercontent.com/3357543/31040200-37c08cea-a539-11e7-9537-9f1d71df53d1.png)

maryx commented on unit_converter/unit_converter/lib/category_route.dart:
```diff
@@ -0,0 +1,68 @@
+                    data[key][i]['conversion'], data[key][i]['description']));
+              }
+              Category category = new Category(key, units);
+              category.onPressed = () => _navigateToConverter(category);
```
@collinjackson I might be doing this wrong. It seems kind of weird to pass in a `category` object into the Category's `onPressed` VoidCallback. But I don't know how else to pass the category I tapped on to `_navigateToConverter()`.

collinjackson replied:
You're right, it is weird. Instead, I would recommend that you move `_navigateToConverter` into your `Category` implementation so you can call it there directly.

maryx replied:
I had thought about that too, but is it weird to be importing `category.dart` into converter_route.dart, and also importing `converter_route.dart` into category.dart?

maryx replied:
Ah, I realized that if I follow your comment below ` replace Category category with List<DropdownMenuItem> units`, I would not need to `import category.dart` into converter_route.dart.

sethladd commented on unit_converter/unit_converter/lib/category_route.dart:
```diff
@@ -0,0 +1,68 @@
+                units.add(new Unit(data[key][i]['name'],
+                    data[key][i]['conversion'], data[key][i]['description']));
+              }
+              Category category = new Category(key, units);
```
You can do this:

`var category = new Category(key, units);`

maryx replied:
Why don't we want to explicitly define the type (Category)? I would've thought this makes the code easier to read. Same for the `decoder` in your comment below.

collinjackson replied:
I think the idea is that omitting optional types is more readable for devs who are coming from languages like JavaScript or Swift?

Personally, I like using types but I hack on the framework and have gotten used to them. :)

maryx replied:
What would be best to teach in the course? `var category = ` or `Category category = `? Thanks!

maryx replied:
Changed to `var category`.

sethladd commented on unit_converter/unit_converter/lib/category_route.dart:
```diff
@@ -0,0 +1,68 @@
+        future: DefaultAssetBundle.of(context).loadString('assets/units.json'),
+        builder: (context, snapshot) {
+          if (snapshot != null && snapshot.data != null) {
+            final JsonDecoder decoder = const JsonDecoder();
```
you can do this:

`final decoder = const JsonDecoder()`

maryx replied:
Done.

sethladd commented on unit_converter/unit_converter/lib/converter_route.dart:
```diff
@@ -0,0 +1,102 @@
+      ),
+    );
+
+    Widget description = new Container(
```
`var description = ...`

maryx replied:
Done.

maryx commented:
Another general question about comments at https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo:
/// triple-slash comments are for documentation for public classes/functions. What about commenting for consts, imports, and global functions? E.g.
```// You can read about packages here: https://flutter.io/using-packages/
import 'package:flutter/material.dart';```

I'm trying to add extra comments and not fully sure when I should use // vs ///.

collinjackson commented on unit_converter/unit_converter/lib/category.dart:
```diff
@@ -0,0 +1,40 @@
+class Category extends StatelessWidget {
+  final String name;
+  final List<Unit> units;
+  VoidCallback onPressed;
```
You can't do this, you'll get an analyzer warning. `StatelessWidget` must be immutable with all fields being `final`.

maryx replied:
Thanks!

collinjackson commented on unit_converter/unit_converter/lib/converter_route.dart:
```diff
@@ -0,0 +1,102 @@
+import 'unit.dart';
+
+class ConverterPage extends StatefulWidget {
+  final Category category;
```
If you want to have a `Category` be model class that isn't a widget, you can do that, but I don't think it should be a widget that also functions as a model class.

In this case, you may want to replace `Category category` with `List<DropdownMenuItem> units` since that's the actual data you need to build the category page.

maryx replied:
Thanks, this makes sense.

collinjackson commented:
I think you should mostly by using // in this app. /// comments are when you want to document something so it can be used outside of its current package and have it show up on dartdocs, whereas // comments are intended for maintainers of the current package (i.e. the app developer)

I'll take a look at https://stackoverflow.com/questions/46530527/how-should-i-customize-dropdownbuttons-and-dropdownmenuitems-in-flutter for your other question.

maryx commented:
Thanks for the comments. Updated code. Ready for more review. Dropdown customization (thanks for answering the Stackoverflow question!) will be in a new PR.

@sethladd @collinjackson I changed various variables to var instead of having a type for them. While it does map to the Javascript idea of `var`, it feels un-intuitive and strange to me when we are required to have types in other parts of the Flutter code. In JS, you don't have to define the return type of a function or its parameters (`function foo(bar, baz) {}`) whereas the same thing in Dart would have to be `void foo(String bar, int baz) {}`. I vote that we define variables with their type in the code instead of using `var`, and explicitly mention in code comments and the script that the student is allowed to use `var`, but they should be reminded that it's still a typed object. What do you think?

sethladd commented:
> What do you think?

We should follow the Dart style and usage guide: https://www.dartlang.org/guides/language/effective-dart/usage#consider-omitting-the-types-for-local-variables

In Dart 2.0, var becomes "infer", instead of "dynamic". So you can look at "var" and think "the compiler will infer the type of this variable".

We strongly encourage new Dart code that is shown to new developers to follow our style and usage guides. In the case of local variables, var is preferred.

For class fields, you can and should use a type annotation. For example:

```
class Foo {
  String bar;
}
```

For any "surface area" of your API, you should use type annotations. Surface area is defined as "the API that a consumer of the API will read and reason about". Your example of `void foo(String bar, int baz)` is great, explicit type annotations are used. This improves tooling tremendously.

For local variables, often the type of the variable is obvious or can be inferred easily.

```
void foo(String bar, int baz) {
  var counter = 0; // obvious this is an int
  var collection = []; // obvious this is a list
  String name = doThing(); // not obvious what doThing returns, so it's OK to use an explicit type annotation here
  var sum = addTwoNumbers(a,b); // pretty obvious this returns a number, OK to use var
}
```

maryx commented:
@sethladd Thanks for posting the guide! This is definitely something we should mention in the video too; I've added it in the comments too so students can learn the style.
I've gone through and updated the types in the code.
Ready for review again.

sethladd commented:
fwiw, it might not be super important to explicitly teach the style guide in the videos. Just show the code as conforming to the style guide, so it becomes implicit: this is how you do it :)

sethladd commented on unit_converter/unit_converter/lib/converter_route.dart:
```diff
@@ -0,0 +1,199 @@
+  String _updateConversion() {
+    if (_inputValue != null) {
+      // Consider omitting the types for local variables. For more details, see
+      // https://www.dartlang.org/guides/language/effective-dart/usage#consider-omitting-the-types-for-local-variables
```
fwiw I don't think it's super important to include these comments. We'll just code with the style guide in mind.

collinjackson commented:
I agree with Seth, and don't have anything else to add.

maryx commented:
Noted about not explicitly teaching the style guide. I think it's still useful to leave a link in the comments in case a student is interested in learning more about it (easy to find).

maryx commented:
Removed duplicate comments and made the comment more generic.

sethladd commented on unit_converter/static_unit_converter/lib/category.dart:
```diff
@@ -0,0 +1,32 @@
+
+class Category {
+  // Builds a row that shows unit category information
+  static Widget buildFromData(String category, Map<String, dynamic> baseUnit) {
```
Does this have to be in a class? Dart has top-level functions.

maryx replied:
No, it doesn't have to be in a class. I created the Category class because later on in the interactive unit converter, we do need to use classes to keep track of some variables, so this helps with the transition to that version.

However, now that you've mentioned top level functions existing, I think it would be beneficial to introduce this concept in the class, so I'll move the `buildFromData` function into a top level function.

sethladd commented on unit_converter/static_unit_converter/lib/main.dart:
```diff
@@ -1,12 +1,17 @@
-import 'dart:core';
+// You can read about packages here: https://flutter.io/using-packages/
```
Every file needs a LICENSE preamble.

maryx replied:
Added.

sethladd commented on unit_converter/static_unit_converter/lib/unit.dart:
```diff
@@ -0,0 +1,42 @@
+import 'package:flutter/material.dart';
```
missing LICENSE preamble. global comment for any .dart file or any file that can accept a comment

maryx replied:
Added.

maryx commented:
Ready for more review.

sethladd commented on unit_converter/static_unit_converter/lib/main.dart:
```diff
@@ -102,85 +110,71 @@ class UnitList extends StatelessWidget {
-            ),
+// Dart allows top level functions
+// Builds a row that shows unit category information
+Widget buildCategory(String category, Map<String, dynamic> baseUnit) {
```
curious, why is baseUnit's keys marked as dynamic?

maryx replied:
The data is:
```
    {
      "name": "Meter",
      "conversion": 1.0,
      "description": "The length of the path traveled by the light in vacuum during a time interval of 1/299,792,458 of a second.",
      "base_unit": true
    },
```
So, the map is from String -> String/double/bool. It doesn't map from String -> a single type.
Do you know of a better way to denote this?

sethladd replied:
Ah, I see. This map is before we convert it to a Dart object.

sethladd commented on unit_converter/unit_converter/lib/unit.dart:
```diff
@@ -0,0 +1,11 @@
+// found in the LICENSE file.
+
+class Unit {
+  String name;
```
could these be final?

maryx replied:
Yep, updated.

sethladd commented on unit_converter/unit_converter/test/widget_test.dart:
```diff
@@ -0,0 +1,29 @@
+// This is a basic Flutter widget test.
```
please add a license preamble here, too

maryx replied:
Added.

sethladd commented:
lgtm
